### PR TITLE
[FIX] web_editor: prevent editing of header

### PR DESCRIPTION
**Pull request by reth-odoo**
## Issue

The way the live HTML editor for html_field works means the header (and other nodes that should only appear once) are removed and the content inserted inside the body.
This results in values being edited without the user being aware.

For emails, this results in any email using 'style' in the header to break on GMail and other clients that do not support 'style' outside of the header.

## Fix

When an html_field is passed a value containing a header the html_field appears readonly and the user is able to toggle the code view. Which enables them to edit the 'raw' HTML and see changes while preventing the presentation view from affecting the database value

task - 3035163

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by thiryr** [comment]
On this last approach which adds a computed field in the back-end. I get about 0.5 seconds for 10k records, 1.5 for 20k.

As far as I'm aware this shouldn't be computed when it's not used either way, but you have the figures if I'm wrong. In that case, we can probably improve that by reducing the margin around which to check from [:500] to [:300] or even [:100] for the data that are currently problematic, or by improving the regex so that it fails early (at the cost of potentially flagging emails that were ok if the regex is too strict).

File: addons/mass_mailing/models/mail_mail.py [hunk 1]
```diff
@@ -24,6 +24,14 @@ def create(self, values_list):
                 mail.mailing_trace_ids.write({'message_id': mail.message_id})
         return mails
 
+    def _compute_contains_uneditable_html(self):
+        processed = self.env['mail.mail']
+        for mail in self:
+            if mail.mailing_id:
+                mail.contains_uneditable_html = True
+                processed += mail
+        super(MailMail, self-processed)._compute_contains_uneditable_html()
```
Comment (std-odoo) [inline addons/mass_mailing/models/mail_mail.py:33]: a bit more readable like this
```py
mailing_mails = self.filtered(lambda mail: mail.mailing_id)
mailing_mails.contains_uneditable_html = True
super(MailMail, self - mailing_mails)._compute_contains_uneditable_html()
```

**Review by std-odoo — COMMENTED** [review]
@thiryr Hi :) 

Don't like the approach but I already did that type of crap because of the web editor :/

Only one comment, otherwise LGTM (if we keep this approach)

**Comment by thiryr** [comment]
force-pushed my first version :p
-> will also add some tests for this and retarget to 16.0

Hi @dmo-odoo I also add toggling the code view in these cases here since I needed to touch the XML anyway but I guess maybe you prefer we just set it readonly? let me know

File: addons/web_editor/static/src/js/backend/html_field.js [hunk 2]
```diff
@@ -74,6 +74,12 @@ export class HtmlFieldWysiwygAdapterComponent extends ComponentAdapter {
 
 export class HtmlField extends Component {
     setup() {
+        // readonly if nodes in <head> as the head will be removed on insertion in the DOM
```
Comment (std-odoo) [inline addons/web_editor/static/src/js/backend/html_field.js:77]: can we add a test ? (so we can well describe the issue in the test docstring)

File: addons/web_editor/static/src/js/backend/html_field.js [hunk 3]
```diff
@@ -74,6 +74,12 @@ export class HtmlFieldWysiwygAdapterComponent extends ComponentAdapter {
...
+        // this prevents the field from overwriting the value with invalid HTML when relevant
+        const domParser = new DOMParser();
+        const parsedOriginal = domParser.parseFromString(this.props.value || '', 'text/html');
+        this.containsComplexHTML = parsedOriginal.head.innerHTML.trim() ? true : false;
```
Comment (std-odoo) [inline addons/web_editor/static/src/js/backend/html_field.js:81]:
```suggestion
        this.containsComplexHTML = !!parsedOriginal.head.innerHTML.trim();
```

File: addons/web_editor/static/src/js/backend/html_field.js [hunk 4]
```diff
@@ -307,23 +313,25 @@ export class HtmlField extends Component {
...
-        } else {
-            this.wysiwyg.odooEditor.observerActive('toggleCodeView');
+        if (this.wysiwyg) {
+            this.wysiwyg.odooEditor.observerUnactive('toggleCodeView');
+            if (this.state.showCodeView) {
+                this.wysiwyg.odooEditor.toolbarHide();
+                const value = this.wysiwyg.getValue();
+                this.props.update(value);
+            }
+            else {
```
Comment (std-odoo) [inline addons/web_editor/static/src/js/backend/html_field.js:327]: my linter is not happy about that 
```suggestion
            } else {
```

File: addons/web_editor/static/src/js/backend/html_field.js [hunk 5]
```diff
@@ -307,23 +313,25 @@ export class HtmlField extends Component {
...
+                this.wysiwyg.odooEditor.observerActive('toggleCodeView');
+            }
+        }
+        if (!this.state.showCodeView) {
             const $codeview = $(this.codeViewRef.el);
             const value = $codeview.val();
             this.props.update(value);
-
```
Comment (std-odoo) [inline addons/web_editor/static/src/js/backend/html_field.js:326]: useless dif (same below)

**Review by std-odoo — COMMENTED** [review]
@thiryr Hi :) 

Thanks for your work :) 

A have a few comments

Cheers

File: addons/web_editor/static/src/js/backend/html_field.js [hunk 6]
```diff
@@ -307,23 +313,25 @@ export class HtmlField extends Component {
...
+            this.wysiwyg.odooEditor.observerUnactive('toggleCodeView');
+            if (this.state.showCodeView) {
+                this.wysiwyg.odooEditor.toolbarHide();
+                const value = this.wysiwyg.getValue();
+                this.props.update(value);
+            }
+            else {
+                this.wysiwyg.odooEditor.observerActive('toggleCodeView');
+            }
+        }
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/backend/html_field.js:330]: I might be missing something but how is this:
```js
if (this.wysiwyg) {
    this.wysiwyg.odooEditor.observerUnactive('toggleCodeView');
    if (this.state.showCodeView) {
        this.wysiwyg.odooEditor.toolbarHide();
        const value = this.wysiwyg.getValue();
          this.props.update(value);
     } else {
          this.wysiwyg.odooEditor.observerActive('toggleCodeView');
     }
}
```
Different than:
```js
if (this.wysiwyg && this.state.showCodeView) {
    this.wysiwyg.odooEditor.observerUnactive('toggleCodeView');
    this.wysiwyg.odooEditor.toolbarHide();
    const value = this.wysiwyg.getValue();
    this.props.update(value);
    this.wysiwyg.odooEditor.observerActive('toggleCodeView');
}
```

File: addons/web_editor/static/src/js/backend/html_field.xml [hunk 7]
```diff
@@ -34,6 +29,13 @@
                 </span>
             </t>
         </t>
+        <t t-if="state.showCodeView || containsComplexHTML">
+            <div t-ref="codeViewButton" id="codeview-btn-group" class="btn-group" t-on-click="() => this.toggleCodeView()">
+                <button class="o_codeview_btn btn btn-primary">
+                    <i class="fa fa-code" />
+                </button>
+            </div>
+        </t>
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/backend/html_field.xml:38]: I'm confused. Why was this moved around ?

File: addons/web_editor/static/src/js/backend/html_field.js (hunk 6 above)
Comment (dmo-odoo) [reply addons/web_editor/static/src/js/backend/html_field.js:329]: I was confused because I feared that we might never reactivate the observer. If that is not the case then this is fine.

File: addons/web_editor/static/src/js/backend/html_field.xml (hunk 7 above)
Comment (dmo-odoo) [reply addons/web_editor/static/src/js/backend/html_field.xml:38]: Oh yes right I get it.

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-ff

**Comment by thiryr** [comment]
Pesky console errors...

It's an issue with our knowledge override of the field, I've added a fix in enterprise. It's pretty straightforward.

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-ff
